core/trino-spi/src/main/java/io/trino/spi/block/DecimalAggregationAccumulatorBlock.java

```diff
@@ -0,0 +1,330 @@
+
+    private final long sizeInBytes;
+    private final long retainedSizeInBytes;
+    private final int numberOfNoOverflowValuesPerEntry;
+    private final int maxOffset;
+
+    public DecimalAggregationAccumulatorBlock(int positionCount, Optional<boolean[]> valueIsNull, long[] values, Optional<long[]> overflow, int numberOfNoOverflowValuesPerEntry)
+    {
+        this(0, positionCount, valueIsNull.orElse(null), values, overflow.orElse(null), numberOfNoOverflowValuesPerEntry);
+    }
+
+    DecimalAggregationAccumulatorBlock(int positionOffset, int positionCount, boolean[] valueIsNull, long[] values, long[] overflow, int numberOfNoOverflowValuesPerEntry)
+    {
+        if (positionOffset < 0) {
+            throw new IllegalArgumentException("positionOffset is negative");
+        }
+        this.positionOffset = positionOffset;
+        if (positionCount < 0) {
+            throw new IllegalArgumentException("positionCount is negative");
+        }
+        this.positionCount = positionCount;
+
+        this.numberOfNoOverflowValuesPerEntry = numberOfNoOverflowValuesPerEntry;
+
+        // offset used for overflow
+        this.maxOffset = 8 * numberOfNoOverflowValuesPerEntry;
+
+        if (values.length - (positionOffset * numberOfNoOverflowValuesPerEntry) < positionCount * numberOfNoOverflowValuesPerEntry) {
+            throw new IllegalArgumentException("values length is less than positionCount");
+        }
+        this.values = values;
+
+        if (valueIsNull != null && valueIsNull.length - positionOffset < positionCount) {
+            throw new IllegalArgumentException("isNull length is less than positionCount");
+        }
+        this.valueIsNull = valueIsNull;
+
+        if (overflow != null && overflow.length - positionOffset < positionCount) {
+            throw new IllegalArgumentException("overflow length is less than positionCount");
+        }
+        this.overflow = overflow;
+
+        sizeInBytes = (numberOfNoOverflowValuesPerEntry + Byte.BYTES + Long.BYTES) * (long) positionCount;
+        retainedSizeInBytes = INSTANCE_SIZE + sizeOf(valueIsNull) + sizeOf(values) + sizeOf(overflow);
+    }
+
+    @Override
+    public int getSliceLength(int position)
+    {
+        return getEntrySizeInBytes();
+    }
+
+    @Override
+    public long getSizeInBytes()
+    {
+        return sizeInBytes;
+    }
+
+    @Override
+    public long getRegionSizeInBytes(int position, int length)
+    {
+        return getEntrySizeInBytes() * (long) length;
+    }
+
+    @Override
+    public long getPositionsSizeInBytes(boolean[] positions)
+    {
+        return getEntrySizeInBytes() * (long) countUsedPositions(positions);
+    }
+
+    @Override
+    public long getRetainedSizeInBytes()
+    {
+        return retainedSizeInBytes;
+    }
+
+    @Override
+    public long getEstimatedDataSizeForStats(int position)
+    {
+        return isNull(position) ? 0 : getEntrySizeInBytes();
+    }
+
+    private int getEntrySizeInBytes()
```

Comment:
move private method below public methods

nit: you could extract this as `final` field